### PR TITLE
8275056: Virtualize G1CardSet containers over heap region 

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

can I have reviews for this change that virtualizes `G1CardSet` "regions" over a heap region, allowing the use of multiple "`G1CardSet` card regions" across a single heap region?

I.e. `HeapRegionRemSet`, which is the interface to a region's card set, simply uses multiple indexes for the remembered set of a single source heap region if necessary. E.g. on a 128MB region, heap region 0's cards would be stored as (what I call) "card region" indexes 0..3 as appropriate in its `_card_set`.

When retrieving the values, the appropriate retransformation is done (during `HeapRegionRemSet::iterate_for_merge()`).

Assigning `HeapRegionRemSet` to handle all this multiplexing required some move of the `G1CardSet::iterate_for_merge` method to `HeapRegionRemSet`, which is why there are more changes than expected.

One change I would like to have opinions on is storing the amount of card regions per region into `G1CardSetConfiguration`, maybe it is better to put this into `HeapRegionRemSet` - but I did not want to start a `HeapRegionRemSetConfiguration` (maybe also put the cached values introduced in the `split_card` optimization there as well?).

This allows unlimited actual heap region size. Currently set to 512MB (what we would set ergonomically if on a 1 TB heap), but that's just a random number basically.
Feel free to suggest a different maximum heap region size if any. We could also keep the ergonomics use a smaller heap region size (e.g. 32M as before).

There is also a CSR to look at.

Testing: tier1-5, some perf testing on region sizes up to 512M with slight improvements in specjbb2015 with larger region sizes.

Thanks,
Thomas

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275056](https://bugs.openjdk.java.net/browse/JDK-8275056): Virtualize G1CardSet containers over heap region

### Reviewers
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to 216504de117e3cb2c63b1c50cac19d4dc263afdf
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to b644a849f6e982fcccc2ecf1ee71a93dffabeddd

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6059/head:pull/6059` \
`$ git checkout pull/6059`

Update a local copy of the PR: \
`$ git checkout pull/6059` \
`$ git pull https://git.openjdk.java.net/jdk pull/6059/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6059`

View PR using the GUI difftool: \
`$ git pr show -t 6059`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6059.diff">https://git.openjdk.java.net/jdk/pull/6059.diff</a>

</details>

### CODE REVIEW

**Comment by tschatzl** [comment]
With the change for JDK-8276548 / PR#6230 now in (I merged it for your convenience), this change works now :) There has been no change in this PR, but the processing of "Full" cards has been wrong with region virtualization.

File: src/hotspot/share/gc/g1/heapRegionRemSet.inline.hpp [hunk 1]
```diff
@@ -55,10 +55,64 @@ void HeapRegionRemSet::set_state_complete() {
...
+    _log_card_regions_per_region(log_card_regions_per_region),
+    _card_regions_per_region_mask((1 << log_card_regions_per_region) - 1),
+    _log_card_region_size(log_card_region_size) {
+  }
+
+  void do_cardsetptr(uint region_idx, size_t num_occupied, G1CardSet::CardSetPtr card_set) override {
+    CardOrRanges<Closure> cl(_iter,
+                             region_idx >> _log_card_regions_per_region,
+                             (region_idx & _card_regions_per_region_mask) << _log_card_region_size);
+    _card_set->iterate_cards_or_ranges_in_container(card_set, cl);
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/heapRegionRemSet.inline.hpp:105]: `region_idx` is actually `card_region_idx`, right? If so, `card_region_idx >> _log_card_regions_per_region` ==> `region_idx` (the second arg of `G1ContainerCardsOrRanges` constructor) makes sense to me.

File: src/hotspot/share/gc/g1/g1CardSet.cpp [hunk 2]
```diff
@@ -44,36 +44,54 @@
...
                          G1RemSetArrayOfCardsEntries,                               /* num_cards_in_array */
                          (double)G1RemSetCoarsenHowlBitmapToHowlFullPercent / 100,  /* cards_in_bitmap_threshold_percent */
                          G1RemSetHowlNumBuckets,                                    /* num_buckets_in_howl */
                          (double)G1RemSetCoarsenHowlToFullPercent / 100,            /* cards_in_howl_threshold_percent */
-                         (uint)HeapRegion::CardsPerRegion)                          /* max_cards_in_cardset */
-                         { }
+                         (uint)HeapRegion::CardsPerRegion,                          /* max_cards_in_cardset */
+                         default_log2_card_region_per_region())                     /* log2_card_region_per_region */
+{
+  assert((_log2_card_region_per_heap_region + _log2_card_region_size) == (uint)HeapRegion::LogCardsPerRegion,
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1CardSet.cpp:67]: I suggest `_log2_card_region_size -> _log2_cards_per_card_region`; then the assertion becomes more consistent, IMO.
`assert((_log2_card_region_per_heap_region + _log2_cards_per_card_region) == (uint)HeapRegion::LogCardsPerRegion)`, which, in math form, is
`(card_region / heap_region) x (cards / card_region) == cards / (heap_)region`

**Review by albertnetymk — APPROVED** [review]
Just some minor comments/suggestions.

File: src/hotspot/share/gc/g1/heapRegionRemSet.inline.hpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/heapRegionRemSet.inline.hpp:105]: Yes, that is the unchanged "region" that the hash table for the containers sees, which is the `card_region_idx` at this level.

File: src/hotspot/share/gc/g1/g1CardSet.cpp (hunk 2 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CardSet.cpp]: Implemented. I'll push in around 7-8 hours unless somebody objects.

**Comment by tschatzl** [comment]
Thanks @albertnetymk @kstefanj for your reviews.

/integrate

**Pull request closed without merging**